drivers/pio/sunxi-pio.c

```diff
@@ -14,56 +15,49 @@
 #include <drivers/pio.h>
 #include <drivers/pio/sunxi-pio.h>
 
-/* This line goes in the header. */
-#define SUNXI_PIO_PIN(port, index) ((port) << 5 | ((index) & 0x1f))
+#define MAX_PORTS         8
+#define MAX_PINS_PER_PORT 32
 
-#define MAX_PORTS                  8
-#define MAX_PINS_PER_PORT          32
+#define PIN_INDEX(pin)    ((pin) & BITMASK(0, 5))
+#define PIN_PORT(pin)     ((pin) >> 5)
 
-#define PIN_INDEX(pin)             ((pin) & BITMASK(0, 5))
-#define PIN_PORT(pin)              ((pin) >> 5)
-
-#define PIN_MODE(mode)             ((mode) & BITMASK(0, 3))
+#define PIN_MODE(mode)    ((mode) & BITMASK(0, 3))
 
 #define CONFIG_REG(pin) \
 	(0x0000 + PIN_PORT(pin) * 0x24 + (PIN_INDEX(pin) >> 3) * 4)
 #define DATA_REG(pin) \
 	(0x0010 + PIN_PORT(pin) * 0x24)
-#define DRIVE_REG(pin) \
-	(0x0014 + PIN_PORT(pin) * 0x24 + (PIN_INDEX(pin) >> 4) * 4)
-#define PULL_REG(pin) \
-	(0x001c + PIN_PORT(pin) * 0x24 + (PIN_INDEX(pin) >> 4) * 4)
 
 #define CONFIG_OFFSET(pin) \
 	((PIN_INDEX(pin) & BITMASK(0, 3)) * 4)
-#define DRIVE_OFFSET(pin) \
-	((PIN_INDEX(pin) & BITMASK(0, 4)) * 2)
-#define PULL_OFFSET(pin) \
-	((PIN_INDEX(pin) & BITMASK(0, 4)) * 2)
-
-#define MODE_MASK 0x7
 
-#define INT_CONFIG_REG(pin) \
-	(0x0200 + PIN_PORT(pin) * 0x20 + (PIN_INDEX(pin) >> 3) * 4)
 #define INT_CONTROL_REG(pin) \
 	(0x0210 + PIN_PORT(pin) * 0x20)
 #define INT_STATUS_REG(pin) \
 	(0x0214 + PIN_PORT(pin) * 0x20)
-#define INT_DEBOUNCE_REG(pin) \
-	(0x0218 + PIN_PORT(pin) * 0x20)
 
-#define INT_CONFIG_OFFSET(pin) \
-	((PIN_INDEX(pin) & BITMASK(0, 3)) * 4)
+/**
+ * Tracks the availability of the I/O ports. A value of 0 indicates an
+ * available port, and a value of 1 indicates an occupied port.
+ */
+static uint8_t ports;
+
+static bool
+sunxi_pio_get_port_status(uint8_t port)
+{
+	return !bitmap_get((uintptr_t)&ports, port);
+}
 
 static bool
 sunxi_pio_read_pin(struct device *dev, uint8_t pin)
 {
-	return !!mmio_read32(dev->regs + DATA_REG(pin), BIT(PIN_INDEX(pin)));
+	return mmio_read32(dev->regs + DATA_REG(pin) + BIT(PIN_INDEX(pin)));
 }
 
 static int
 sunxi_pio_set_mode(struct device *dev, uint8_t pin, uint8_t mode)
 {
+	/* Set port to unavailable. */
```

Comment:
I don't see where you actually do this.

Reply:
There was a missing `TODO` here, but this has been added in the latest push.